### ISSUE TITLE
ValueError: arange: cannot compute length

### ISSUE THREAD

**Issue by franciskim-yonsei** [welch-lab/MultiVelo#28]
Hi
Thank you for creating an great tool.
When I run the following code, a problem occurs. Can you help me solve it?

Code:
-----------------------------------------------------------------------------
adata_result = mv.recover_dynamics_chrom(ldata_subset,
                                         adata_subset,
                                         max_iter=5,
                                         init_mode="invert",
                                         verbose=False,
                                         parallel=True,
                                         save_plot=False,
                                         rna_only=False,
                                         fit=True,
                                         n_anchors=500,
                                        )
--------------------------------------------------------------------------------

Error:
---------------------------------------------------------------------------
_RemoteTraceback                          Traceback (most recent call last)
_RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/user/envs/scVelo/lib/python3.9/site-packages/joblib/externals/loky/process_executor.py", line 463, in _process_worker
    r = call_item()
  File "/home/user/envs/scVelo/lib/python3.9/site-packages/joblib/externals/loky/process_executor.py", line 291, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/home/user/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py", line 588, in __call__
    return [func(*args, **kwargs)
  File "/home/user/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py", line 588, in 
    return [func(*args, **kwargs)
  File "/home/user/envs/scVelo/lib/python3.9/site-packages/multivelo/dynamical_chrom_func.py", line 2549, in regress_func
    cdc = ChromatinDynamical(c,
  File "/home/user/envs/scVelo/lib/python3.9/site-packages/multivelo/dynamical_chrom_func.py", line 1074, in __init__
    self.initialize_steady_state_params(model_mismatch=self.model
  File "/home/user/envs/scVelo/lib/python3.9/site-packages/multivelo/dynamical_chrom_func.py", line 1386, in initialize_steady_state_params
    for t_sw_1 in np.arange(1, rna_interval-1, 2,
ValueError: arange: cannot compute length
"""

The above exception was the direct cause of the following exception:

ValueError                                Traceback (most recent call last)
Cell In[385], line 1
----> 1 adata_result = mv.recover_dynamics_chrom(ldata_subset,
      2                                          adata_subset,
      3                                          max_iter=5,
      4                                          init_mode="invert",
      5                                          verbose=False,
      6                                          parallel=True,
      7                                          save_plot=False,
      8                                          rna_only=False,
      9                                          fit=True,
     10                                          n_anchors=500,
     11                                         )

File [~/envs/scVelo/lib/python3.9/site-packages/multivelo/dynamical_chrom_func.py:3107](https://file+.vscode-resource.vscode-cdn.net/home/user/ojh/scVelo/~/envs/scVelo/lib/python3.9/site-packages/multivelo/dynamical_chrom_func.py:3107), in recover_dynamics_chrom(adata_rna, adata_atac, gene_list, max_iter, init_mode, model_to_run, verbose, plot, parallel, n_jobs, save_plot, plot_dir, rna_only, fit, fit_decoupling, extra_color_key, embedding, n_anchors, k_dist, thresh_multiplier, weight_c, outlier, n_pcs, n_neighbors, fig_size, point_size, partial, direction, rescale_u, alpha, beta, gamma, t_sw)
   3104 verbose = False
   3105 plot = False
-> 3107 res = Parallel(n_jobs=n_jobs, backend='loky', verbose=verb)(
   3108     delayed(func_to_call)(
   3109         c_mat[:, i],
   3110         u_mat[:, i],
   3111         s_mat[:, i],
   3112         model_to_run[i] if m_per_g else model_to_run,
   3113         max_iter,
   3114         init_mode,
   3115         global_pdist,
   3116         embed_coord,
   3117         rna_conn,
   3118         verbose,
   3119         plot,
   3120         save_plot,
   3121         plot_dir,
   3122         fit_args,
   3123         gene_list[i],
   3124         partial[i] if p_per_g else partial,
   3125         direction[i] if d_per_g else direction,
   3126         rna_only,
   3127         fit,
   3128         fit_decoupling,
   3129         extra_color,
   3130         ru[i] if isinstance(ru, (list, np.ndarray)) else ru,
   3131         alpha[i] if isinstance(alpha, (list, np.ndarray))
   3132         else alpha,
   3133         beta[i] if isinstance(beta, (list, np.ndarray))
   3134         else beta,
   3135         gamma[i] if isinstance(gamma, (list, np.ndarray))
   3136         else gamma,
   3137         t_sw[i] if isinstance(t_sw, (list, np.ndarray)) else t_sw)
   3138     for i in gene_indices)
   3140 for i, r in zip(gene_indices, res):
   3141     (loss, model, direct_out, parameters, initial_exp,
   3142      time, state, velocity, likelihood, anchors) = r

File [~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:1944](https://file+.vscode-resource.vscode-cdn.net/home/user/ojh/scVelo/~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:1944), in Parallel.__call__(self, iterable)
   1938 # The first item from the output is blank, but it makes the interpreter
   1939 # progress until it enters the Try/Except block of the generator and
   1940 # reach the first `yield` statement. This starts the aynchronous
   1941 # dispatch of the tasks to the workers.
   1942 next(output)
-> 1944 return output if self.return_generator else list(output)

File [~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:1587](https://file+.vscode-resource.vscode-cdn.net/home/user/ojh/scVelo/~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:1587), in Parallel._get_outputs(self, iterator, pre_dispatch)
   1584     yield
   1586     with self._backend.retrieval_context():
-> 1587         yield from self._retrieve()
   1589 except GeneratorExit:
   1590     # The generator has been garbage collected before being fully
   1591     # consumed. This aborts the remaining tasks if possible and warn
   1592     # the user if necessary.
   1593     self._exception = True

File [~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:1691](https://file+.vscode-resource.vscode-cdn.net/home/user/ojh/scVelo/~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:1691), in Parallel._retrieve(self)
   1684 while self._wait_retrieval():
   1685 
   1686     # If the callback thread of a worker has signaled that its task
   1687     # triggered an exception, or if the retrieval loop has raised an
   1688     # exception (e.g. `GeneratorExit`), exit the loop and surface the
   1689     # worker traceback.
   1690     if self._aborting:
-> 1691         self._raise_error_fast()
   1692         break
   1694     # If the next job is not ready for retrieval yet, we just wait for
   1695     # async callbacks to progress.

File [~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:1726](https://file+.vscode-resource.vscode-cdn.net/home/user/ojh/scVelo/~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:1726), in Parallel._raise_error_fast(self)
   1722 # If this error job exists, immediatly raise the error by
   1723 # calling get_result. This job might not exists if abort has been
   1724 # called directly or if the generator is gc'ed.
   1725 if error_job is not None:
-> 1726     error_job.get_result(self.timeout)

File [~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:735](https://file+.vscode-resource.vscode-cdn.net/home/user/ojh/scVelo/~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:735), in BatchCompletionCallBack.get_result(self, timeout)
    729 backend = self.parallel._backend
    731 if backend.supports_retrieve_callback:
    732     # We assume that the result has already been retrieved by the
    733     # callback thread, and is stored internally. It's just waiting to
    734     # be returned.
--> 735     return self._return_or_raise()
    737 # For other backends, the main thread needs to run the retrieval step.
    738 try:

File [~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:753](https://file+.vscode-resource.vscode-cdn.net/home/user/ojh/scVelo/~/envs/scVelo/lib/python3.9/site-packages/joblib/parallel.py:753), in BatchCompletionCallBack._return_or_raise(self)
    751 try:
    752     if self.status == TASK_ERROR:
--> 753         raise self._result
    754     return self._result
    755 finally:

ValueError: arange: cannot compute length

-------------------------------------------------------------------------------------------------------------

I don't know much about the code. Therefore, I tried asking ChatGPT.

Answer:

------------------------------------------------------
The error message you provided seems to have originated from the arange function in numpy. This function generates numbers at a certain interval from a start value to an end value. The error occurred because the supplied parameters are not appropriate, and it cannot calculate the length.

In the code, it seems that the problem occurred in the part for t_sw_1 in np.arange(1, rna_interval-1, 2,. For np.arange(1, rna_interval-1, 2), the rna_interval-1 needs to be greater than 1. If the rna_interval is less than or equal to 2, arange can't generate a valid sequence, leading to an error.

To solve this problem, you should first check the value of rna_interval. Ensure that rna_interval is set to an appropriate value, and adjust this value if necessary. This is especially important if rna_interval represents a certain attribute of the data.

It would also be good to review the entire code as there may be issues in other parts of the code. Particularly, verify that the inputs to the recover_dynamics_chrom function are correct. Make sure the parameters of the function are working as expected, and that each variable has the right range and format.

Additionally, since this error occurred during a parallel processing operation, there may be issues in the part of the code managing parallel processing. Check if your parallel processing settings are appropriate and adjust them if necessary.

For more specific assistance, additional information on the input variable values for the recover_dynamics_chrom function or how the mv.recover_dynamics_chrom function was implemented would be helpful. Providing this information will allow me to provide more detailed advice.

------------------------------------------------------------------------------

**Comment by danielee0707** [comment]
Hi,
What's the version of MultiVelo that you're using? It could be related to issue #5.

**Comment by franciskim-yonsei** [comment]
Name: multivelo
Version: 0.1.3
Summary: Multi-omic extension of single-cell RNA velocity
Home-page: https://github.com/welch-lab/MultiVelo
Author: Chen Li
Author-email: chlseven@umich.edu
License: BSD 3-Clause License
Location: [/home/user/envs/scVelo/lib/python3.9/site-packages](https://file+.vscode-resource.vscode-cdn.net/home/user/envs/scVelo/lib/python3.9/site-packages)
Requires: anndata, h5py, ipywidgets, joblib, loompy, matplotlib, numba, numpy, pandas, scanpy, scikit-learn, scipy, scvelo, seaborn, tqdm, umap-learn
Required-by:

The version of Multivelo installed in my Python is 0.1.3

is it new released version?

**Comment by danielee0707** [comment]
Yes, that seems to be the latest version. Do you know which gene was causing the error at the end? Can you plot its phase portrait with scvelo.pl.scatter?

**Comment by franciskim-yonsei** [comment]
The previous error disappeared. So, I tried again by creating a new one, but a similar error occurred. It stopped at variable number 8544 out of 19516 variables. Therefore, I looked for adata_rna.var.iloc[8543] and found the result:

Accession: ENSMUSG00000112294.1
Chromosome: 10
End: 123199418
Start: 123196695
Strand: +
Gene_count_corr: -0.01

I plotted the values for Gm4129 using scvelo.pl.scatter, and the result is as follows.


![Gm4129](https://github.com/welch-lab/MultiVelo/assets/127166114/bf81a96c-3690-4429-8e9a-9809dc54989a)

**Comment by danielee0707** [comment]
Thanks for sharing the plot. It seems like this gene doesn't have high-quality unspliced and spliced counts. Was any filtering performed on the unspliced and spliced counts during preprocessing steps prior to calling the recover dynamics function?
In the meantime, it is safe to continue the run without this gene.

**Comment by franciskim-yonsei** [comment]
Following your advice on filtering, it worked out well

Thanks!!

**Comment by danielee0707** [comment]
Glad it worked out. In case there are more ambiguous errors, our team has recently added more robust logging functions to the GitHub version of the package, and it should be helpful for debugging purposes. It will be bundled into a new release in the near future.
Closing the issue due to completion.

**Comment by Kyung-TaeLee** [comment]
Hi, I know that this issue is closed but I am stuck with the same problem. Version of Multivelo package I am using is 0.1.5 so I believe it is the latest version. I did not filter the cells or genes since I already annotated cells with cell types and I only used the cells with cell type labels. Top 1000 variables genes were used for downstream analysis and there are 546 cells. After running recover_dynamics_chrom, I also got this "ValueError: arange: cannot compute length". Expression plot of the gene where the process stopped is shown below.

<img width="764" height="473" alt="Image" src="https://github.com/user-attachments/assets/d97343d3-af3a-4175-a26e-1c862fdbc96b" />

Is there a way to systematically remove the genes that are causing the issues altogether or should I just run it and remove the genes causing the error by manually? Thank you for your help in advance.
